File: python/2.py
# --- Day 2: Dive! ---

from sys import argv

if __name__ == '__main__':
    # open file sys argv 1 or default to 1.txt
    input1 = open(argv[1] if len(argv) > 1 else '2.txt', 'r')
    data = [(i.split(" ")[0], int(i.split(" ")[1])) for i in input1]

    pos, aim, dth = 0, 0, 0

    for op, n in data:
        if op == "forward":
            pos += n
            dth += aim * n
        elif op == "down":
            aim += n
        elif op == "up":
            aim -= n

    print(f"Part 1: {pos * aim}")
    print(f"Part 2: {pos * dth}")

File: golang/2.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func main() {
	file, err := os.Open("../inputs/2.txt")

	if err != nil {
		panic(err)
	}

	defer file.Close()
	scanner := bufio.NewScanner(file)
	var pos, aim, dth uint64 = 0, 0, 0

	for scanner.Scan() {
		line := strings.Split(scanner.Text(), " ")
		op := line[0]
		val, err := strconv.ParseUint(line[1], 10, 64)

		if err != nil {
			panic(err)
		}

		switch op {
		case "down":
			aim += val
		case "up":
			aim -= val
		case "forward":
			pos += val
			dth += aim * val
		}
	}

	fmt.Println("Part 1:", pos*aim)
	fmt.Println("Part 2:", pos*dth)
}
